### PR TITLE
Fix ACL errors for newly created and pre-existing blobs

### LINKED ISSUE #944: make_public fails due to ACL being NULL on buckets/blobs that are pre-loaded into the docker container
I get an ACL error when using make_public on data that is used to initialize the docker container.

this is the sample python code:

```python
    os.environ["STORAGE_EMULATOR_HOST"] = "http://0.0.0.0:4443"

    client = storage.Client(
        credentials=AnonymousCredentials(),
        project="test-project",
    )

    # initial bucket/file for docker image
    bucket_name = "sample-bucket"
    blob_name = "some_file.txt"

    # test
    bucket = client.get_bucket(bucket_name)
    blob = bucket.blob(blob_name)
    blob.make_public()  #             <<<<< this errors on None not being iterable
    blob.make_private()  #             <<<<< this too
```

this gives `TypeError: 'NoneType' object is not iterable` in google/cloud/storage/acl.py", line 398
(as far as I can tell this is somewhere in `response.go` from aclListResponse to getAccessControlsListFromObject, where it somehow returns {"items":null} instead of {"items":[]} ??)

Doing the same thing with a bucket and blob created from code works (when #943 is applied)

---

_Originally posted by @marcguidancenl in https://github.com/fsouza/fake-gcs-server/issues/941#issuecomment-1269647236_

### LINKED ISSUE #945: make_public and make_private don't actually update the ACL
I noticed that the ACL doesn't actually get changed with make_public and make_private. Sample code (only works after applying #943, otherwise you get an error, see #941):

```python
    os.environ["STORAGE_EMULATOR_HOST"] = "http://0.0.0.0:4443"

    client = storage.Client(
        credentials=AnonymousCredentials(),
        project="test-project",
    )

    # buckets/files to create and test
    upload_bucket_name = "test-bucket-with-globally-unique-name"
    upload_blob_name = "test-blob-upload.svg"
    upload_blob_file_path = "./image.svg"

    # initialize
    try:
        bucket = client.bucket(upload_bucket_name)
        bucket.storage_class = storage.constants.STANDARD_STORAGE_CLASS
        client.create_bucket(bucket, location="EU", retry=None)
    except:
        pass
    bucket = client.get_bucket(upload_bucket_name)
    blob = bucket.blob(upload_blob_name)
    blob.upload_from_filename(upload_blob_file_path, retry=None)

    # test
    bucket = client.get_bucket(upload_bucket_name)
    blob = bucket.blob(upload_blob_name)
    print(list(blob.acl))
    blob.make_public()
    print(list(blob.acl))  #            <<<<< didn't change ACL entities!
    blob.make_private()
    print(list(blob.acl))
```

this lists the same ACL three times - initial ACL, public ACL, and private ACL:

```
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]
```

I think the second one should at least have 'allUsers' + 'READER'.

---

_Originally posted by @marcguidancenl in https://github.com/fsouza/fake-gcs-server/issues/941#issuecomment-1269647236_

### PR DESCRIPTION
**Pull request by RachitSharma2001**
This fixes #944 and fixes #945.

The following two python snippets no longer crash and instead give the correct output. The first one is for a pre-existing blob (thus fixing #944):

```
os.environ["STORAGE_EMULATOR_HOST"] = "http://0.0.0.0:4443"

client = storage.Client(
    credentials=AnonymousCredentials(),
    project="test-project",
)

# initial bucket/file for docker image
bucket_name = "sample-bucket"
blob_name = "some_file.txt"

# test
bucket = client.get_bucket(bucket_name)
blob = bucket.blob(blob_name)
print(list(blob.acl))
blob.make_public() 
print(list(blob.acl))
blob.make_private()
print(list(blob.acl))
```
The outputs:
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}, {'entity': 'allUsers', 'role': 'READER'}]
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]

The second snippet shows how for a newly created blob, the acl's also update when make_public() and make_private() are called:
```
os.environ["STORAGE_EMULATOR_HOST"] = "http://0.0.0.0:4443"

client = storage.Client(
    credentials=AnonymousCredentials(),
    project="test-project",
)

# buckets/files to create and test
upload_bucket_name = "test-bucket-with-globally-unique-name"
upload_blob_name = "test-blob-upload.svg"
upload_blob_file_path = "./image.svg"

# initialize
try:
    bucket = client.bucket(upload_bucket_name)
    bucket.storage_class = storage.constants.STANDARD_STORAGE_CLASS
    client.create_bucket(bucket, location="EU", retry=None)
except:
    pass
bucket = client.get_bucket(upload_bucket_name)
blob = bucket.blob(upload_blob_name)
blob.upload_from_filename(upload_blob_file_path, retry=None)

# test
bucket = client.get_bucket(upload_bucket_name)
blob = bucket.blob(upload_blob_name)
print(list(blob.acl))
blob.make_public()
print(list(blob.acl)) 
blob.make_private()
print(list(blob.acl))
```
The outputs:
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}, {'entity': 'allUsers', 'role': 'READER'}]
[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]

Explanation of my implementation:

For fixing #944:
Within main.go, when it reads all the existing files within the bucket, it previously never set an ACL for any of these objects. I assumed that for these pre-existing blobs, the ACL for each of them would just be `[{'entity': 'projectOwner-test-project', 'role': 'OWNER'}]`.

For fixing #945:
From looking at how the python API updates the ACL's, I found that rather than sending a POST to the endpoint `/b/{bucketName}/o/{objectName:.+}/acl`, it instead sends a PATCH request to the endpoint `/b/{bucketName}/o/{objectName:.+}`. 

Thus, I needed to update the `patchObject` method within `fakestorage/object.go` to detect if new ACL's are passed in and update the object's ACL if so.

### CODE REVIEW

File: fakestorage/object.go [hunk 1]
```diff
@@ -934,17 +934,31 @@ func (s *Server) patchObject(r *http.Request) jsonResponse {
...
 	bucketName := vars["bucketName"]
 	objectName := vars["objectName"]
-	var metadata struct {
+
+	type acls struct {
+		Entity string
+		Role   string
+	}
+
+	var dataInBody struct {
```
Comment (fsouza) [inline fakestorage/object.go:943]: nit: can we call this `payload`?
```suggestion
	var payload struct {
```

File: fakestorage/object.go [hunk 2]
```diff
@@ -934,17 +934,31 @@ func (s *Server) patchObject(r *http.Request) jsonResponse {
...
 		}
 	}
-	backendObj, err := s.backend.PatchObject(bucketName, objectName, metadata.Metadata)
+	backendObj, err := s.backend.PatchObject(bucketName, objectName, dataInBody.Metadata)
+	if len(dataInBody.Acl) > 0 {
+		backendObj.ACL = []storage.ACLRule{}
+		for _, aclData := range dataInBody.Acl {
+			newAcl := storage.ACLRule{Entity: storage.ACLEntity(aclData.Entity), Role: storage.ACLRole(aclData.Role)}
+			backendObj.ACL = append(backendObj.ACL, newAcl)
+		}
```
Comment (fsouza) [inline fakestorage/object.go:960]: Hmm this won't work with the backend filesystem, will it? We're modifying the value in memory and not persisting it. We need to update the PatchObject signature to take something like "attrsToUpdate" and apply changes to other attributes, not just metadata (this is necessary for #1024 too).

File: main.go [hunk 3]
```diff
@@ -14,12 +14,20 @@ import (
...
 	"syscall"
 
+	"cloud.google.com/go/storage"
 	"github.com/fsouza/fake-gcs-server/fakestorage"
 	"github.com/fsouza/fake-gcs-server/internal/checksum"
 	"github.com/fsouza/fake-gcs-server/internal/config"
 	"github.com/sirupsen/logrus"
 )
 
+var DefaultACL = []storage.ACLRule{
```
Comment (fsouza) [inline main.go:24]: This doesn't really need to be a global, let's inline it (I know it means we'll repeat it in tests, but that's OK).

File: fakestorage/object.go (hunk 2 above)
Comment (RachitSharma2001) [reply fakestorage/object.go:960]: Ah good point. I will fix this and push the changes.

Comment (RachitSharma2001) [reply fakestorage/object.go]: I have made the requested change. Let me know if what I did was best practice or if there is a better way.

File: internal/backend/fs.go [hunk 4]
```diff
@@ -310,21 +311,34 @@ func (s *storageFS) DeleteObject(bucketName, objectName string) error {
 	return os.Remove(path)
 }
 
-// PatchObject patches the given object metadata.
-func (s *storageFS) PatchObject(bucketName, objectName string, metadata map[string]string) (StreamingObject, error) {
+// func (s *storageFS) PatchObject(bucketName, objectName string, metadata map[string]string) (StreamingObject, error) {
```
Comment (fsouza) [inline internal/backend/fs.go:314]: can probably delete this line?

File: internal/backend/fs.go [hunk 5]
```diff
@@ -310,21 +311,34 @@ func (s *storageFS) DeleteObject(bucketName, objectName string) error {
...
+		} else if currObjType.Field(i).Name == "Metadata" {
+			if obj.Metadata == nil {
+				obj.Metadata = map[string]string{}
+			}
+			for k, v := range attrsToUpdate.Metadata {
+				obj.Metadata[k] = v
+			}
+		} else {
+			currObjValues.Field(i).Set(newObjValues.Field(i))
+		}
```
Comment (fsouza) [inline internal/backend/fs.go:337]: nit: could we abstract this in a helper function? It's the same in both backends, right?

**Review by fsouza — APPROVED** [review]
Thanks for working on this! Just one nit on generalizing some code, otherwise it's ready to go.

If you prefer, I can take care of it.

**Review by fsouza — APPROVED** [review]
Thank you very much for contributing!

**Pull request merged**